File: frontend/src/api/focus.ts
import request from './index';

// 专注概览响应
export interface FocusOverview {
  today_pomodoro_count: number;
  today_focus_duration: number;  // 秒
  total_pomodoro_count: number;
  total_focus_duration: number;  // 秒
}

// 专注记录
export interface FocusSession {
  id: string;
  user_id: string;
  task_id?: string;
  task_title?: string;
  type: 'pomodoro' | 'stopwatch';
  duration: number;  // 秒
  started_at: string;
  ended_at: string;
  created_at: string;
}

// 专注记录列表响应
export interface FocusSessionsResponse {
  sessions: FocusSession[];
  total: number;
  page: number;
  page_size: number;
}

// 创建专注记录请求
export interface FocusSessionCreateRequest {
  task_id?: string;
  type: 'pomodoro' | 'stopwatch';
  duration: number;
  started_at: string;
  ended_at: string;
}

// 获取专注概览
export const getFocusOverview = (): Promise<FocusOverview> => {
  return request.get('/focus/overview');
};

// 获取专注记录列表
export const getFocusSessions = (params?: {
  page?: number;
  page_size?: number;
  start_date?: string;
  end_date?: string;
}): Promise<FocusSessionsResponse> => {
  return request.get('/focus/sessions', { params });
};

// 创建专注记录
export const createFocusSession = (data: FocusSessionCreateRequest): Promise<FocusSession> => {
  return request.post('/focus/sessions', data);
};

// 删除专注记录
export const deleteFocusSession = (id: string): Promise<{ success: boolean }> => {
  return request.delete(`/focus/sessions/${id}`);
};

File: backend/routes/focus.py
# -*- coding: utf-8 -*-

from fastapi import APIRouter, HTTPException, Depends, Query
from pydantic import BaseModel, Field
from typing import Optional

from middleware.jwt_middleware import get_current_user
from database.dao.focus_dao import focus_dao, FocusSessionTaskNotFoundError

router = APIRouter(prefix="/api/focus", tags=["focus"])


class FocusSessionCreate(BaseModel):
    """创建专注记录请求体"""
    task_id: Optional[str] = None
    type: str = "pomodoro"        # "pomodoro" / "stopwatch"
    duration: int = 0              # 秒
    started_at: str = ""
    ended_at: str = ""
    # 固定小写 UUID 避免 SQLite/MySQL 的大小写、重音及尾空格比较差异。
    # 旧客户端可省略或传 null；非规范键拒绝，不静默归一化/合并记录。
    client_session_id: Optional[str] = Field(
        default=None,
        min_length=36,
        max_length=36,
        pattern=r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
        description="Canonical lowercase hyphenated UUID; reuse unchanged for retries",
    )


@router.get("/overview")
async def get_focus_overview(current_user_id: str = Depends(get_current_user)):
    """获取专注概览"""
    try:
        return focus_dao.get_overview(current_user_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f'获取专注概览失败: {str(e)}')


@router.get("/capabilities")
async def get_focus_capabilities(current_user_id: str = Depends(get_current_user)):
    """返回客户端可用的专注接口能力。"""
    return {"client_session_id": True}


@router.get("/sessions")
async def get_focus_sessions(
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    start_date: Optional[str] = None,
    end_date: Optional[str] = None,
    current_user_id: str = Depends(get_current_user)
):
    """获取专注记录列表"""
    try:
        return focus_dao.get_sessions(current_user_id, page, page_size, start_date, end_date)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f'获取专注记录失败: {str(e)}')


@router.post("/sessions")
async def create_focus_session(
    session: FocusSessionCreate,
    current_user_id: str = Depends(get_current_user)
):
    """创建专注记录"""
    try:
        return focus_dao.create_session(current_user_id, session.dict())
    except FocusSessionTaskNotFoundError:
        raise HTTPException(status_code=404, detail="Task not found")
    except Exception as e:
        raise HTTPException(status_code=500, detail=f'创建专注记录失败: {str(e)}')


@router.delete("/sessions/{session_id}")
async def delete_focus_session(
    session_id: str,
    current_user_id: str = Depends(get_current_user)
):
    """删除专注记录"""
    try:
        success = focus_dao.delete_session(current_user_id, session_id)
        if not success:
            raise HTTPException(status_code=404, detail="Session not found")
        return {"success": True}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f'删除专注记录失败: {str(e)}')
